src/session_impl.cpp

```diff
@@ -3732,11 +3732,56 @@ namespace aux {
 	}
 
 	namespace {
-		bool last_optimistic_unchoke_cmp(torrent_peer const* const l
-			, torrent_peer const* const r)
+		struct opt_unchoke_candidate
 		{
-			return l->last_optimistically_unchoked
-				< r->last_optimistically_unchoked;
+			opt_unchoke_candidate(boost::shared_ptr<peer_connection> const* tp
+				, std::vector<boost::shared_ptr<plugin>>* ps)
+				: peer(tp)
+				, plugins(ps)
+				, ext_priority(std::numeric_limits<uint64_t>::max())
+			{}
+
+			uint64_t get_ext_priority() const
+			{
+#ifndef TORRENT_DISABLE_EXTENSIONS
+				if (plugins)
+				{
+					for (auto& e : *plugins)
+					{
+						uint64_t const priority = e->get_unchoke_priority(peer_connection_handle(*peer));
+						ext_priority = (std::min)(priority, ext_priority);
+					}
+					plugins = nullptr;
+				}
+#endif
+				return ext_priority;
+			}
+
+			boost::shared_ptr<peer_connection> const* peer;
+
+		private:
+			// these are mutable because comparison functors passed to std::partial_sort
+			// are not supposed to modify the elements they are sorting. Here the mutation
+			// being applied is idempotent so it should not pose a problem.
+			mutable std::vector<boost::shared_ptr<plugin>>* plugins;
```

Comment:
sorry for nit-picking. wouldn't it make more sense for this pointer to live in the comparator function object? Now it's duplicated lots of times. I realize you also use this as an indicator whether the ext_priority has been computed, would it be unreasonable to instead use a special value to indicate this, say UINT64_MAX?